src/uwtools/drivers/driver.py

```diff
@@ -25,19 +27,27 @@ class Driver(ABC):
     """
 
     def __init__(
-        self, config: Optional[Path] = None, dry_run: bool = False, batch: bool = False
+        self,
+        config: Optional[Path] = None,
+        dry_run: bool = False,
+        batch: bool = False,
+        cycle: Optional[datetime] = None,
     ) -> None:
         """
         A component driver.
 
         :param config: Path to config file (read stdin if missing or None).
         :param dry_run: Run in dry-run mode?
         :param batch: Run component via the batch system?
+        :param cycle: The synoptic time.
         """
         self._config = YAMLConfig(config=config)
         self._dry_run = dry_run
         self._batch = batch
         self._config.dereference()
+        self._config.dereference(
+            context={**({"cycle": cycle} if cycle else {}), **self._config.data}
+        )
```

Comment:
When dereferencing a driver config, start with `cycle`, if it's available, then layer on the content from the YAML config. If the user, for whatever reason, provides a YAML config that defines `cycle`, the user value will take precedence.

As explained elsewhere, and given a template expression like
```
{{ rootdir }}/{{ cycle.strftime('%Y%m%d') }}/{{ cycle.strftime('%H') }}/run/namelist.wps
```
this solves a problem whereby the first dereference in `Driver` would fail to render the expression because, although it had a `rootdir` value in the YAML config, it did not have a `cycle` value; and the second dereference in the driver constructor would fail because, although _it_ has `cycle`, it would not have `rootdir`.

For example, given `ungrib.yaml`
```
ungrib:
  execution:
    batchargs:
      cores: 1
      walltime: 00:01:00
    executable: /path/to/ungrib.exe
  gfs_file: /path/to/dir/gfs.t12z.pgrb2.0p25.f000
  run_dir: "{{ rootdir }}/{{ cycle.strftime('%Y%m%d') }}/{{ cycle.strftime('%H') }}/run"
  vtable: /path/to/Vtable.GFS
rootdir: /tmp
platform:
  account: me
  scheduler: slurm
```
Old behavior:
```
% uw ungrib namelist_file --dry-run --cycle 2024-03-22T18 --config ungrib.yaml 
[2024-03-22T23:33:03]     INFO Validating config against internal schema ungrib
[2024-03-22T23:33:03]     INFO 0 UW schema-validation errors found
[2024-03-22T23:33:03]     INFO Validating config against internal schema platform
[2024-03-22T23:33:03]     INFO 0 UW schema-validation errors found
[2024-03-22T23:33:03]     INFO 20240322 18Z ungrib {{ rootdir }}/{{ cycle.strftime('%Y%m%d') }}/{{ cycle.strftime('%H') }}/run/namelist.wps: Initial state: Pending
[2024-03-22T23:33:03]     INFO 20240322 18Z ungrib {{ rootdir }}/{{ cycle.strftime('%Y%m%d') }}/{{ cycle.strftime('%H') }}/run/namelist.wps: Checking requirements
[2024-03-22T23:33:03]     INFO 20240322 18Z ungrib {{ rootdir }}/{{ cycle.strftime('%Y%m%d') }}/{{ cycle.strftime('%H') }}/run/namelist.wps: Requirement(s) ready
[2024-03-22T23:33:03]     INFO 20240322 18Z ungrib {{ rootdir }}/{{ cycle.strftime('%Y%m%d') }}/{{ cycle.strftime('%H') }}/run/namelist.wps: SKIPPING (DRY RUN)
```
New behavior:
```
% uw ungrib namelist_file --dry-run --cycle 2024-03-22T18 --config ungrib.yaml 
[2024-03-22T23:33:26]     INFO Validating config against internal schema ungrib
[2024-03-22T23:33:26]     INFO 0 UW schema-validation errors found
[2024-03-22T23:33:26]     INFO Validating config against internal schema platform
[2024-03-22T23:33:26]     INFO 0 UW schema-validation errors found
[2024-03-22T23:33:26]     INFO 20240322 18Z ungrib /tmp/20240322/18/run/namelist.wps: Initial state: Pending
[2024-03-22T23:33:26]     INFO 20240322 18Z ungrib /tmp/20240322/18/run/namelist.wps: Checking requirements
[2024-03-22T23:33:26]     INFO 20240322 18Z ungrib /tmp/20240322/18/run/namelist.wps: Requirement(s) ready
[2024-03-22T23:33:26]     INFO 20240322 18Z ungrib /tmp/20240322/18/run/namelist.wps: SKIPPING (DRY RUN)
```